// postcss/lib/rule.js
"use strict";

var _prototypeProperties = function (child, staticProps, instanceProps) {
  if (staticProps) Object.defineProperties(child, staticProps);
  if (instanceProps) Object.defineProperties(child.prototype, instanceProps);
};

var _get = function get(object, property, receiver) {
  var desc = Object.getOwnPropertyDescriptor(object, property);

  if (desc === undefined) {
    var parent = Object.getPrototypeOf(object);

    if (parent === null) {
      return undefined;
    } else {
      return get(parent, property, receiver);
    }
  } else if ("value" in desc && desc.writable) {
    return desc.value;
  } else {
    var getter = desc.get;
    if (getter === undefined) {
      return undefined;
    }
    return getter.call(receiver);
  }
};

var _inherits = function (subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
  }
  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {
      value: subClass,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });
  if (superClass) subClass.__proto__ = superClass;
};

var Container = require("./container");
var Declaration = require("./declaration");
var list = require("./list");

// CSS rule like “a { }”
var Rule = (function (Container) {
  function Rule(defaults) {
    this.type = "rule";
    this.nodes = [];
    _get(Object.getPrototypeOf(Rule.prototype), "constructor", this).call(this, defaults);
  }

  _inherits(Rule, Container);

  _prototypeProperties(Rule, null, {
    selectors: {

      // Shortcut to get selectors as array

      get: function () {
        return list.comma(this.selector);
      },
      set: function (values) {
        this.selector = values.join(", ");
      },
      enumerable: true,
      configurable: true
    },
    stringify: {

      // Stringify rule
      value: function stringify(builder) {
        this.stringifyBlock(builder, this.stringifyRaw("selector"));
      },
      writable: true,
      enumerable: true,
      configurable: true
    }
  });

  return Rule;
})(Container);

module.exports = Rule;
